consensus/qbft/core/backlog.go

```diff
@@ -0,0 +1,177 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"github.com/ethereum/go-ethereum/consensus/qbft"
+	"github.com/ethereum/go-ethereum/consensus/qbft/message"
+	"gopkg.in/karalabe/cookiejar.v2/collections/prque"
+)
+
+var (
+	// msgPriority is defined for calculating processing priority to speedup consensus
+	// msgPreprepare > msgCommit > msgPrepare
+	msgPriority = map[uint64]int{
+		message.PreprepareCode: 1,
+		message.CommitCode:     2,
+		message.PrepareCode:    3,
+	}
+)
+
+// checkMessage checks the message state
+// return errInvalidMessage if the message is invalid
+// return errFutureMessage if the message view is larger than current view
+// return errOldMessage if the message view is smaller than current view
+func (c *core) checkMessage(msgCode uint64, view *qbft.View) error {
+	if view == nil || view.Sequence == nil || view.Round == nil {
+		return errInvalidMessage
+	}
+
+	if msgCode == message.RoundChangeCode {
+		if view.Sequence.Cmp(c.currentView().Sequence) > 0 {
+			return errFutureMessage
+		} else if view.Cmp(c.currentView()) < 0 {
+			return errOldMessage
+		}
+		return nil
+	}
+
+	if view.Cmp(c.currentView()) > 0 {
+		return errFutureMessage
+	}
+
+	if view.Cmp(c.currentView()) < 0 {
+		return errOldMessage
+	}
+
+	switch c.state {
+	case StateAcceptRequest:
+		// StateAcceptRequest only accepts msgPreprepare and msgRoundChange
+		// other messages are future messages
+		if msgCode > message.PreprepareCode {
+			return errFutureMessage
+		}
+		return nil
+	case StatePreprepared:
+		// StatePreprepared only accepts msgPrepare and msgRoundChange
+		// message less than msgPrepare are invalid and greater are future messages
+		if msgCode < message.PrepareCode {
+			return errInvalidMessage
+		} else if msgCode > message.PrepareCode {
+			return errFutureMessage
+		}
+		return nil
+	case StatePrepared:
+		// StatePrepared only accepts msgCommit and msgRoundChange
+		// other messages are invalid messages
+		if msgCode < message.CommitCode {
+			return errInvalidMessage
+		}
+		return nil
+	case StateCommitted:
+		// StateCommit rejects all messages other than msgRoundChange
+		return errInvalidMessage
+	}
+	return nil
+}
+
+func (c *core) storeQBFTBacklog(msg message.QBFTMessage) {
+	src := msg.Source()
+	logger := c.logger.New("from", src, "state", c.state)
+
+	if src == c.Address() {
+		logger.Warn("Backlog from self")
+		return
+	}
+
+	logger.Trace("Store future message")
+
+	c.backlogsMu.Lock()
+	defer c.backlogsMu.Unlock()
+
+	logger.Debug("Retrieving backlog queue", "for", src, "backlogs_size", len(c.backlogs))
+	backlog := c.backlogs[src]
+	if backlog == nil {
+		backlog = prque.New()
```

Comment:
`c.backlogs[src] = backlog` could be done under the `if` statement for clarity

Reply:
Moved `c.backlogs[src] = backlog` to under `if` statement